server/src/internalClusterTest/java/org/elasticsearch/search/basic/TransportSearchFailuresIT.java

```diff
@@ -121,4 +139,76 @@ private XContentBuilder source(String id, String nameValue, int age) throws IOEx
                 .field("multi", multi.toString())
                 .endObject();
     }
+
+    public void testRetryQueryPhase() throws Exception {
+        assertAcked(prepareCreate("test"));
+
+        NumShards numShards = getNumShards("test");
+        int numPrimaries = numShards.numPrimaries;
+
+        for (int i = 0; i < 10; i++) {
+            index(client(), Integer.toString(i), "test", i);
+        }
+        refresh("test");
+
+        SearchResponse searchResponse = client().prepareSearch("test")
+            .setSource(new SearchSourceBuilder().query(new FailingQueryBuilder()))
+            .setSearchType(SearchType.QUERY_RETRY_THEN_FETCH)
+            .get();
+        assertThat(searchResponse.getTotalShards(), equalTo(numPrimaries));
+        assertThat(searchResponse.getSuccessfulShards(), equalTo(numPrimaries));
+        assertThat(searchResponse.getFailedShards(), equalTo(0));
+    }
+
+    public static class FailingQueryPlugin extends Plugin implements SearchPlugin {
+        @Override
+        public List<QuerySpec<?>> getQueries() {
+            return List.of(new QuerySpec<>(FailingQueryBuilder.NAME, FailingQueryBuilder::new, parserContext -> new FailingQueryBuilder()));
+        }
+    }
+
+    // TODO(jtibs): this is a big hack and not a solid testing strategy
```

Comment:
We could add another test where an index is created with shard allocation filtering that would prevent the node being allocated, execute the search, spin up a node that satisfies the allocation requirements and wait to get the search results back.